diff --git a/browser/base/content/browser.js b/browser/base/content/browser.js
--- a/browser/base/content/browser.js
+++ b/browser/base/content/browser.js
@@ -5448,49 +5448,54 @@ function asyncOpenWebPanel(event)
 
 /*
  * - [ Dependencies ] ---------------------------------------------------------
  *  utilityOverlay.js:
  *    - gatherTextUnder
  */
 
 /**
- * Extracts linkNode and href for the current click target.
+ * Extracts linkNode and href for the passed in element.
  *
- * @param event
- *        The click event.
+ * @param aElement
+ *        The element to extract the information from.
  * @return [href, linkNode].
  *
- * @note linkNode will be null if the click wasn't on an anchor
- *       element (or XLink).
+ * @note linkNode will be null if element wasn't an anchor element.
  */
-function hrefAndLinkNodeForClickEvent(event)
+function hrefAndLinkNodeForElement(aElement)
 {
   function isHTMLLink(aNode)
   {
     // Be consistent with what nsContextMenu.js does.
     return ((aNode instanceof HTMLAnchorElement && aNode.href) ||
             (aNode instanceof HTMLAreaElement && aNode.href) ||
             aNode instanceof HTMLLinkElement);
   }
 
-  let node = event.target;
+  let node = aElement;
   while (node && !isHTMLLink(node)) {
     node = node.parentNode;
   }
 
   if (node)
     return [node.href, node];
 
-  // If there is no linkNode, try simple XLink.
+  // If there is no linkNode, try MathML link or simple XLink.
   let href, baseURI;
-  node = event.target;
+  node = aElement;
   while (node && !href) {
     if (node.nodeType == Node.ELEMENT_NODE) {
-      href = node.getAttributeNS("http://www.w3.org/1999/xlink", "href");
+      // MathML3 allows to define a href attribute, that has priority on XLinks.
+      if (node.namespaceURI == "http://www.w3.org/1998/Math/MathML" &&
+          node.hasAttribute("href")) {
+        href = node.getAttribute("href");
+      } else {
+        href = node.getAttributeNS("http://www.w3.org/1999/xlink", "href");
+      }
       if (href)
         baseURI = node.baseURI;
     }
     node = node.parentNode;
   }
 
   // In case of XLink, we don't return the node we got href from since
   // callers expect <a>-like elements.
@@ -5506,17 +5511,17 @@ function hrefAndLinkNodeForClickEvent(ev
  *        Whether the event comes from a web panel.
  * @note default event is prevented if the click is handled.
  */
 function contentAreaClick(event, isPanelClick)
 {
   if (!event.isTrusted || event.getPreventDefault() || event.button == 2)
     return true;
 
-  let [href, linkNode] = hrefAndLinkNodeForClickEvent(event);
+  let [href, linkNode] = hrefAndLinkNodeForElement(event.target);
   if (!href) {
     // Not a link, handle middle mouse navigation.
     if (event.button == 1 &&
         gPrefService.getBoolPref("middlemouse.contentLoadURL") &&
         !gPrefService.getBoolPref("general.autoScroll")) {
       middleMousePaste(event);
       event.preventDefault();
     }
diff --git a/browser/base/content/nsContextMenu.js b/browser/base/content/nsContextMenu.js
--- a/browser/base/content/nsContextMenu.js
+++ b/browser/base/content/nsContextMenu.js
@@ -414,16 +414,20 @@ nsContextMenu.prototype = {
       this.setItemAttr("context-media-showcontrols", "disabled", hasError);
       this.setItemAttr("context-media-hidecontrols", "disabled", hasError);
       if (this.onVideo)
         this.setItemAttr("context-video-fullscreen",  "disabled", hasError);
     }
     this.showItem("context-media-sep-commands",  onMedia);
   },
 
+  get onLink() !!this.link,
+  get onMailtoLink() this.linkProtocol === "mailto",
+  get onSaveableLink() this.onLink && this.isLinkSaveable(this.link),
+
   // Set various context menu attributes based on the state of the world.
   setTarget: function (aNode, aRangeParent, aRangeOffset) {
     const xulNS = "http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul";
     if (aNode.namespaceURI == xulNS ||
         aNode.nodeType == Node.DOCUMENT_NODE ||
         this.isTargetAFormControl(aNode)) {
       this.shouldDisplay = false;
       return;
@@ -435,19 +439,16 @@ nsContextMenu.prototype = {
     this.onCompletedImage  = false;
     this.onStandaloneImage = false;
     this.onCanvas          = false;
     this.onVideo           = false;
     this.onAudio           = false;
     this.onTextInput       = false;
     this.onKeywordField    = false;
     this.mediaURL          = "";
-    this.onLink            = false;
-    this.onMailtoLink      = false;
-    this.onSaveableLink    = false;
     this.link              = null;
     this.linkURL           = "";
     this.linkURI           = null;
     this.linkProtocol      = "";
     this.onMathML          = false;
     this.inFrame           = false;
     this.hasBGImage        = false;
     this.bgImageURL        = "";
@@ -535,35 +536,25 @@ nsContextMenu.prototype = {
     }
 
     // Second, bubble out, looking for items of interest that can have childen.
     // Always pick the innermost link, background image, etc.
     const XMLNS = "http://www.w3.org/XML/1998/namespace";
     var elem = this.target;
     while (elem) {
       if (elem.nodeType == Node.ELEMENT_NODE) {
-        // Link?
-        if (!this.onLink &&
-            // Be consistent with what hrefAndLinkNodeForClickEvent
-            // does in browser.js
-             ((elem instanceof HTMLAnchorElement && elem.href) ||
-              (elem instanceof HTMLAreaElement && elem.href) ||
-              elem instanceof HTMLLinkElement ||
-              elem.getAttributeNS("http://www.w3.org/1999/xlink", "type") == "simple")) {
-
-          // Target is a link or a descendant of a link.
-          this.onLink = true;
-
-          // Remember corresponding element.
-          this.link = elem;
-          this.linkURL = this.getLinkURL();
-          this.linkURI = this.getLinkURI();
-          this.linkProtocol = this.getLinkProtocol();
-          this.onMailtoLink = (this.linkProtocol == "mailto");
-          this.onSaveableLink = this.isLinkSaveable( this.link );
+        if (!this.onLink) {
+          let [href, linkNode] = hrefAndLinkNodeForElement(elem);
+          if (href) {
+            // Target is a link or a descendant of a link.
+            this.link = elem;
+            this.linkURL = href;
+            this.linkURI = makeURI(href);
+            this.linkProtocol = this.linkURI.scheme;
+          }
         }
 
         // Background image?  Don't bother if we've already found a
         // background image further down the hierarchy.  Otherwise,
         // we look for the computed background-image style.
         if (!this.hasBGImage &&
             !this._hasMultipleBGImages) {
           let bgImgUrl;
@@ -1140,50 +1131,21 @@ nsContextMenu.prototype = {
       node.setAttribute(attr.nodeName, attr.nodeValue);
     }
 
     // Voila!
     return node;
   },
 
   // Generate fully qualified URL for clicked-on link.
-  getLinkURL: function() {
-    var href = this.link.href;  
-    if (href)
-      return href;
+  getLinkURL: function() this.linkURL,
 
-    href = this.link.getAttributeNS("http://www.w3.org/1999/xlink",
-                                    "href");
+  getLinkURI: function() this.linkURI,
 
-    if (!href || !href.match(/\S/)) {
-      // Without this we try to save as the current doc,
-      // for example, HTML case also throws if empty
-      throw "Empty href";
-    }
-
-    return makeURLAbsolute(this.link.baseURI, href);
-  },
-
-  getLinkURI: function() {
-    try {
-      return makeURI(this.linkURL);
-    }
-    catch (ex) {
-     // e.g. empty URL string
-    }
-
-    return null;
-  },
-
-  getLinkProtocol: function() {
-    if (this.linkURI)
-      return this.linkURI.scheme; // can be |undefined|
-
-    return null;
-  },
+  getLinkProtocol: function() this.linkProtocol, // can be |undefined|
 
   // Get text of link.
   linkText: function() {
     var text = gatherTextUnder(this.link);
     if (!text || !text.match(/\S/)) {
       text = this.link.getAttribute("title");
       if (!text || !text.match(/\S/)) {
         text = this.link.getAttribute("alt");
diff --git a/browser/base/content/test/browser_contentAreaClick.js b/browser/base/content/test/browser_contentAreaClick.js
--- a/browser/base/content/test/browser_contentAreaClick.js
+++ b/browser/base/content/test/browser_contentAreaClick.js
@@ -48,28 +48,30 @@
 
 let gTests = [
 
   {
     desc: "Simple left click",
     setup: function() {},
     clean: function() {},
     event: {},
-    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink" ],
+    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink",
+               "mathhref", "mathhrefnested", "mathxlinkandhref" ],
     expectedInvokedMethods: [],
     preventDefault: false,
   },
 
   {
     desc: "Ctrl/Cmd left click",
     setup: function() {},
     clean: function() {},
     event: { ctrlKey: true,
              metaKey: true },
-    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink" ],
+    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink",
+               "mathhref", "mathhrefnested", "mathxlinkandhref" ],
     expectedInvokedMethods: [ "urlSecurityCheck", "openLinkIn" ],
     preventDefault: true,
   },
 
   // The next test was once handling feedService.forcePreview().  Now it should
   // just be like Alt click.
   {
     desc: "Shift+Alt left click",
@@ -83,27 +85,29 @@ let gTests = [
   },
 
   {
     desc: "Shift+Alt left click on XLinks",
     setup: function() {},
     clean: function() {},
     event: { shiftKey: true,
              altKey: true },
-    targets: [ "mathxlink", "svgxlink"],
+    targets: [ "mathxlink", "svgxlink",
+               "mathhref", "mathhrefnested", "mathxlinkandhref" ],
     expectedInvokedMethods: [ "saveURL" ],
     preventDefault: true,
   },
 
   {
     desc: "Shift click",
     setup: function() {},
     clean: function() {},
     event: { shiftKey: true },
-    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink" ],
+    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink",
+               "mathhref", "mathhrefnested", "mathxlinkandhref" ],
     expectedInvokedMethods: [ "urlSecurityCheck", "openLinkIn" ],
     preventDefault: true,
   },
 
   {
     desc: "Alt click",
     setup: function() {},
     clean: function() {},
@@ -113,17 +117,18 @@ let gTests = [
     preventDefault: true,
   },
 
   {
     desc: "Alt click on XLinks",
     setup: function() {},
     clean: function() {},
     event: { altKey: true },
-    targets: [ "mathxlink", "svgxlink" ],
+    targets: [ "mathxlink", "svgxlink",
+               "mathhref", "mathhrefnested", "mathxlinkandhref" ],
     expectedInvokedMethods: [ "saveURL" ],
     preventDefault: true,
   },
 
   {
     desc: "Panel click",
     setup: function() {},
     clean: function() {},
@@ -133,33 +138,35 @@ let gTests = [
     preventDefault: true,
   },
 
   {
     desc: "Simple middle click opentab",
     setup: function() {},
     clean: function() {},
     event: { button: 1 },
-    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink" ],
+    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink",
+               "mathhref", "mathhrefnested", "mathxlinkandhref" ],
     expectedInvokedMethods: [ "urlSecurityCheck", "openLinkIn" ],
     preventDefault: true,
   },
 
   {
     desc: "Simple middle click openwin",
     setup: function() {
       gPrefService.setBoolPref("browser.tabs.opentabfor.middleclick", false);
     },
     clean: function() {
       try {
         gPrefService.clearUserPref("browser.tabs.opentabfor.middleclick");
       } catch(ex) {}
     },
     event: { button: 1 },
-    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink" ],
+    targets: [ "commonlink", "mathxlink", "svgxlink", "maplink",
+               "mathhref", "mathhrefnested", "mathxlinkandhref" ],
     expectedInvokedMethods: [ "urlSecurityCheck", "openLinkIn" ],
     preventDefault: true,
   },
 
   {
     desc: "Middle mouse paste",
     setup: function() {
       gPrefService.setBoolPref("middlemouse.contentLoadURL", true);
@@ -254,18 +261,23 @@ let gClickHandler = {
     executeSoon(runNextTest);
   }
 }
 
 // Wraps around the methods' replacement mock function.
 function wrapperMethod(aInvokedMethods, aMethodName) {
   return function () {
     aInvokedMethods.push(aMethodName);
+
     // At least getShortcutOrURI requires to return url that is the first param.
-    return arguments[0];
+    let maybeURL = arguments[0];
+    if (/^http/.test(maybeURL)) {
+      is(maybeURL, "http://mochi.test/moz/", "The correct URL was requested");
+    }
+    return maybeURL;
   }
 }
 
 function setupTestBrowserWindow() {
   // Steal click events and don't propagate them.
   gTestWin.addEventListener("click", gClickHandler, true);
 
   // Replace methods.
@@ -278,17 +290,20 @@ function setupTestBrowserWindow() {
   let doc = gTestWin.content.document;
   let mainDiv = doc.createElement("div");
   mainDiv.innerHTML =
     '<p><a id="commonlink" href="http://mochi.test/moz/">Common link</a></p>' +
     '<p><a id="panellink" href="http://mochi.test/moz/">Panel link</a></p>' +
     '<p><a id="emptylink">Empty link</a></p>' +
     '<p><math id="mathxlink" xmlns="http://www.w3.org/1998/Math/MathML" xlink:type="simple" xlink:href="http://mochi.test/moz/"><mtext>MathML XLink</mtext></math></p>' +
     '<p><svg id="svgxlink" xmlns="http://www.w3.org/2000/svg" width="100px" height="50px" version="1.1"><a xlink:type="simple" xlink:href="http://mochi.test/moz/"><text transform="translate(10, 25)">SVG XLink</text></a></svg></p>' +
-    '<p><map name="map" id="map"><area href="http://mochi.test/moz/" shape="rect" coords="0,0,128,128" /></map><img id="maplink" usemap="#map" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIAAAACACAIAAABMXPacAAAABGdBTUEAALGPC%2FxhBQAAAOtJREFUeF7t0IEAAAAAgKD9qRcphAoDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGBgwIAAAT0N51AAAAAASUVORK5CYII%3D"/></p>'
+    '<p><map name="map" id="map"><area href="http://mochi.test/moz/" shape="rect" coords="0,0,128,128" /></map><img id="maplink" usemap="#map" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIAAAACACAIAAABMXPacAAAABGdBTUEAALGPC%2FxhBQAAAOtJREFUeF7t0IEAAAAAgKD9qRcphAoDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGDAgAEDBgwYMGBgwIAAAT0N51AAAAAASUVORK5CYII%3D"/></p>' +
+    '<p><math id="mathhref" xmlns="http://www.w3.org/1998/Math/MathML" href="http://mochi.test/moz/"><mtext>MathML href</mtext></math></p>' +
+    '<p><math id="mathhrefnested" xmlns="http://www.w3.org/1998/Math/MathML"><mtext href="http://mochi.test/moz/">MathML href</mtext></math></p>' +
+    '<p><math id="mathxlinkandhref" xmlns="http://www.w3.org/1998/Math/MathML"><mtext xlink:href="http://mochi.test/wrong_link/" href="http://mochi.test/moz/">MathML href</mtext></math></p>'
   doc.body.appendChild(mainDiv);
 }
 
 function runNextTest() {
   if (!gCurrentTest) {
     gCurrentTest = gTests.shift();
     gCurrentTest.setup();
   }
